File: westcoast/src/main/java/frc/commands/intake/IntakeCollect.java
package frc.commands.intake;

import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.subsystems.IntakeSubsystem;
import frc.subsystems.MagazineSubsystem;
import org.jetbrains.annotations.NotNull;

import java.util.function.BooleanSupplier;

public final class IntakeCollect extends CommandBase {

    private enum State {
        End,
        Full,
        Collecting,
        Rejecting,
        AcceptingToFeeder,
        AcceptingToIndex,
    }

    @NotNull
    private final IntakeSubsystem intake;
    @NotNull
    private final MagazineSubsystem magazine;
    @NotNull
    private final BooleanSupplier ballAcceptanceStrategy;

    private final int rejectTickCount;
    private final int reverseIntakeTickCount;
    private final boolean finishWhenFull;

    private State state = State.Collecting;

    public IntakeCollect(
            @NotNull IntakeSubsystem intake,
            @NotNull MagazineSubsystem magazine,
            @NotNull BooleanSupplier ballAcceptanceStrategy,
            int rejectTickCount,
            int reverseIntakeTickCount,
            boolean finishWhenFull
    ) {
        this.intake = intake;
        this.magazine = magazine;
        this.ballAcceptanceStrategy = ballAcceptanceStrategy;
        this.rejectTickCount = rejectTickCount;
        this.reverseIntakeTickCount = reverseIntakeTickCount;
        this.finishWhenFull = finishWhenFull;
        addRequirements(intake, magazine);
    }

    @Override
    public void initialize() {
        state = State.Collecting;
    }

    @Override
    public void execute() {
        switch (state) {
            case Full:
                full();
                break;
            case Collecting:
                reachReject = false;
                reverseIntakeRunningTime = 0;
                collect();
                break;
            case Rejecting:
                reject();
                break;
            case AcceptingToFeeder:
                acceptToFeeder();
                break;
            case AcceptingToIndex:
                acceptToIndex();
                break;
            case End:
                break;
        }
    }

    private void full() {
        magazine.stopAll();
        intake.stop();
        if (finishWhenFull) {
            state = State.End;
        } else if (!magazine.hasIndexSensorBreak() || !magazine.hasFeederSensorBreak()) {
            state = State.Collecting;
        }
    }

    private void collect() {
        if (!magazine.hasIndexSensorBreak()) {
            intake.start();
            magazine.startIndexMotor();
        } else {
            intake.stop();
            magazine.stopIndexMotor();
            if (ballAcceptanceStrategy.getAsBoolean()) {
                if (!magazine.hasFeederSensorBreak()) {
                    state = State.AcceptingToFeeder;
                } else {
                    state = State.AcceptingToIndex;
                }
            } else {
                state = State.Rejecting;
            }
        }
    }

    private int rejectRunningTime = 0;
    private boolean reachReject = false;

    private void reject() {
        if (magazine.hasRejectSensorBreak()) {
            reachReject = true;
            rejectRunningTime = 0;
        }

        if (!reachReject) {
            magazine.startIndexMotor();
            magazine.startRejectMotor();
        } else if (rejectRunningTime < rejectTickCount) {
            rejectRunningTime++;
            magazine.startIndexMotor();
            magazine.startRejectMotor();
        } else {
            magazine.stopIndexMotor();
            magazine.stopRejectMotor();
            state = State.Collecting;
        }
    }

    private void acceptToFeeder() {
        if (!magazine.hasFeederSensorBreak()) {
            magazine.startIndexMotor();
            magazine.startFeederMotor();
            magazine.reverseRejectMotor();
        } else {
            magazine.stopIndexMotor();
            magazine.stopFeederMotor();
            magazine.stopRejectMotor();
            state = State.Collecting;
        }
    }

    private int reverseIntakeRunningTime = 0;

    private void acceptToIndex() {
        if (reverseIntakeRunningTime < reverseIntakeTickCount) {
            reverseIntakeRunningTime++;
            intake.reverse();
        } else {
            intake.stop();
            state = State.Full;
        }
    }

    @Override
    public void end(boolean interrupted) {
        intake.stop();
        magazine.stopAll();
    }

    @Override
    public boolean isFinished() {
        return state == State.End;
    }
}

File: westcoast/src/main/kotlin/frc/commands/intake/IntakeCollect.kt
package frc.commands.intake

import edu.wpi.first.wpilibj2.command.CommandBase
import frc.subsystems.IntakeSubsystem
import frc.subsystems.MagazineSubsystem

class IntakeCollect(
    private val intake: IntakeSubsystem,
    private val magazine: MagazineSubsystem,
    private val ballAcceptanceStrategy: () -> Boolean,
    private val rejectTickCount: Int,
    private val reverseIntakeTickCount: Int,
    private val finishWhenFull: Boolean
) : CommandBase() {
    private enum class State {
        End,
        Full,
        Collecting,
        Rejecting,
        AcceptingToFeeder,
        AcceptingToIndex
    }

    init {
        addRequirements(intake, magazine)
    }

    private var state = State.Collecting
    override fun initialize() {
        state = State.Collecting
    }

    override fun execute() {
        when (state) {
            State.Full -> full()
            State.Collecting -> {
                reachReject = false
                reverseIntakeRunningTime = 0
                collect()
            }
            State.Rejecting -> reject()
            State.AcceptingToFeeder -> acceptToFeeder()
            State.AcceptingToIndex -> acceptToIndex()
            State.End -> Unit
        }
    }

    private fun full() {
        magazine.stopAll()
        intake.stop()
        state = when {
            finishWhenFull -> State.End
            !magazine.hasIndexSensorBreak() || !magazine.hasFeederSensorBreak() -> State.Collecting
            else -> state
        }
    }

    private fun collect() {
        if (!magazine.hasIndexSensorBreak()) {
            intake.start()
            magazine.startIndexMotor()
        } else {
            intake.stop()
            magazine.stopIndexMotor()
            state = when {
                !ballAcceptanceStrategy() -> State.Rejecting
                !magazine.hasFeederSensorBreak() -> State.AcceptingToFeeder
                else -> State.AcceptingToIndex
            }
        }
    }

    private var rejectRunningTime = 0
    private var reachReject = false
    private fun reject() {
        if (magazine.hasRejectSensorBreak()) {
            reachReject = true
            rejectRunningTime = 0
        }
        when {
            !reachReject -> {
                magazine.startIndexMotor()
                magazine.startRejectMotor()
            }
            rejectRunningTime < rejectTickCount -> {
                rejectRunningTime++
                magazine.startIndexMotor()
                magazine.startRejectMotor()
            }
            else -> {
                magazine.stopIndexMotor()
                magazine.stopRejectMotor()
                state = State.Collecting
            }
        }
    }

    private fun acceptToFeeder() {
        if (!magazine.hasFeederSensorBreak()) {
            magazine.startIndexMotor()
            magazine.startFeederMotor()
            magazine.reverseRejectMotor()
        } else {
            magazine.stopIndexMotor()
            magazine.stopFeederMotor()
            magazine.stopRejectMotor()
            state = State.Collecting
        }
    }

    private var reverseIntakeRunningTime = 0

    private fun acceptToIndex() {
        if (reverseIntakeRunningTime < reverseIntakeTickCount) {
            reverseIntakeRunningTime++
            intake.reverse()
        } else {
            intake.stop()
            state = State.Full
        }
    }

    override fun end(interrupted: Boolean) {
        intake.stop()
        magazine.stopAll()
    }

    override fun isFinished(): Boolean {
        return state == State.End
    }
}